toxcore/tox_api.c

```diff
@@ -282,21 +308,61 @@ void tox_options_set_experimental_disable_dns(Tox_Options *options, bool experim
 {
     options->experimental_disable_dns = experimental_disable_dns;
 }
+bool tox_options_get_experimental_owned_data(const Tox_Options *options)
+{
+    return options->experimental_owned_data;
+}
+void tox_options_set_experimental_owned_data(
+    Tox_Options *options, bool experimental_owned_data)
+{
+    options->experimental_owned_data = experimental_owned_data;
+}
 
 const uint8_t *tox_options_get_savedata_data(const Tox_Options *options)
 {
     return options->savedata_data;
 }
 
-void tox_options_set_savedata_data(Tox_Options *options, const uint8_t *savedata_data, size_t length)
+bool tox_options_set_savedata_data(Tox_Options *options, const uint8_t *savedata_data, size_t length)
 {
-    options->savedata_data = savedata_data;
+    if (!options->experimental_owned_data) {
+        options->savedata_data = savedata_data;
+        options->savedata_length = length;
+        return true;
+    }
+
+    if (options->owned_savedata_data != nullptr) {
+        free(options->owned_savedata_data);
+        options->owned_savedata_data = nullptr;
+    }
+    if (savedata_data == nullptr) {
+        options->savedata_data = nullptr;
+        options->savedata_length = 0;
+        return false;
+    }
+
+    uint8_t *owned_ptr = (uint8_t *)malloc(length);
+    if (owned_ptr != nullptr) {
+        memcpy(owned_ptr, savedata_data, length);
+        options->savedata_data = owned_ptr;
+        options->owned_savedata_data = owned_ptr;
+    } else {
+        options->savedata_data = savedata_data;
+        options->owned_savedata_data = nullptr;
```

Comment:
>  Client code can check whether
allocation succeeded by checking the returned bool. If
experimental_owned_data is false, it will always return true. If set to
true, the return value will be false on allocation failure.

Reply:
Good catch. Also the set-to-nullptr case now returns true.